File: core/src/com/xenoage/zong/commands/core/music/lyric/LyricAdd.java
package com.xenoage.zong.commands.core.music.lyric;

import static com.xenoage.utils.collections.CollectionUtils.addOrNew;
import static com.xenoage.utils.collections.CollectionUtils.removeOrEmpty;
import lombok.AllArgsConstructor;

import com.xenoage.utils.document.command.Command;
import com.xenoage.utils.document.command.Undoability;
import com.xenoage.zong.core.music.chord.Chord;
import com.xenoage.zong.core.music.lyric.Lyric;


/**
 * Adds the given {@link Lyric} to the given {@link Chord}.
 * 
 * @author Andreas Wenger
 */
@AllArgsConstructor public class LyricAdd
	implements Command {
	
	//data
	private Lyric lyric;
	private Chord chord;

	@Override public void execute() {
		chord.setLyrics(addOrNew(chord.getLyrics(), lyric));
	}

	@Override public Undoability getUndoability() {
		return Undoability.Undoable;
	}

	@Override public void undo() {
		chord.setLyrics(removeOrEmpty(chord.getLyrics(), lyric));
	}

}

File: core/src/com/xenoage/zong/commands/core/music/lyric/LyricAdd.kt
package com.xenoage.zong.commands.core.music.lyric

import com.xenoage.utils.collections.addOrNew
import com.xenoage.utils.document.command.UndoableCommand
import com.xenoage.utils.collections.removeOrEmpty
import com.xenoage.zong.core.music.chord.Chord
import com.xenoage.zong.core.music.lyric.Lyric


/**
 * Adds the given [Lyric] to the given [Chord].
 */
class LyricAdd(
		private val lyric: Lyric,
		private val chord: Chord
) : UndoableCommand() {

	override fun execute() {
		chord.lyrics = chord.lyrics.addOrNew(lyric)
	}

	override fun undo() {
		chord.lyrics = chord.lyrics.removeOrEmpty(lyric)
	}

}
